### PR TITLE
Multi-provider testing

### PR DESCRIPTION
**Pull request by kmoe**
Some Terraform provider acceptance tests make use of other Terraform providers. We will refer to these as _multi-provider acceptance tests_, in which a _primary provider_ under test uses resources from _auxiliary providers_. 

For example, consider the following (abbreviated) test configuration string in [resource_kubernetes_persistent_volume_test.go](https://github.com/terraform-providers/terraform-provider-kubernetes/blob/master/kubernetes/resource_kubernetes_persistent_volume_test.go) (terraform-provider-kubernetes):
```
resource "kubernetes_persistent_volume" "test" {
  spec {
    persistent_volume_source {
      aws_elastic_block_store {
        volume_id = "${aws_ebs_volume.test.id}"
      }
    }
  }
}
resource "aws_ebs_volume" "test" {
  size = 10
}
```
Here, the primary `kubernetes` provider uses a resource from the auxiliary `google` provider.

The binary test driver will first attempt to load auxiliary provider binaries stored in `terraform.d` in the provider source code repository itself. If the required provider binary is not found there, it will fall back to the Terraform CLI’s plugin discovery mechanism.

### CODE REVIEW

File: helper.go [hunk 1]
```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
...
+// the provider source code directory root in order to use this feature.
+func symlinkAuxiliaryProviders(pluginDir string) error {
+	providerRootDir := os.Getenv("TF_ACC_PROVIDER_ROOT_DIR")
+	if providerRootDir == "" {
+		// common case; assume intentional and do not log
+		return nil
+	}
+
+	_, err := os.Stat(filepath.Join(providerRootDir, "terraform.d", "plugins"))
+	if err != nil {
```
Comment (appilon) [inline helper.go]: ensure not exist with `if os.IsNotExist(err) {`

File: helper.go [hunk 2]
```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
...
+		fmt.Printf("No terraform.d/plugins directory found: continuing. Unset TF_ACC_PROVIDER_ROOT_DIR or supply provider binaries in terraform.d/plugins/$GOOS_$GOARCH to disable this message.")
+		return nil
+	}
+
+	auxiliaryProviderDir := filepath.Join(providerRootDir, "terraform.d", "plugins", runtime.GOOS+"_"+runtime.GOARCH)
+
+	// If we can't os.Stat() terraform.d/plugins/$GOOS_$GOARCH, however,
+	// assume the omission was unintentional, and error.
+	_, err = os.Stat(auxiliaryProviderDir)
+	if err != nil {
```
Comment (appilon) [inline helper.go]: ensure not exist with `if os.IsNotExist(err) {`

File: helper.go (hunk 1 above)
Comment (appilon) [reply helper.go]: can throw in an `else if err !=nil { return fmt.Errorf("Unexpected error: %s", err) }`

File: helper.go [hunk 3]
```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
...
+		_, err := os.Stat(providerSymlinkPath)
+		if err == nil {
+			continue
+		}
+
+		// if filename ends in .zip, assume it is a zip and extract it
+		// otherwise assume it is a provider binary
+		if providerNameExt == ".zip" {
+			_, err = os.Stat(providerPath)
+			if err == nil {
```
Comment (appilon) [inline helper.go]: I would check `os.IsNotExist(err)`, do the decompress, `else if err != nil` return unexpected error, omit `else`.

**Review by appilon — CHANGES_REQUESTED** [review]
Just minor nitpicks over the errors return from `os.Stat`, likely it's overkill. Other than that lookin great, really interested in this.

File: helper.go (hunk 2 above)
Comment (appilon) [reply helper.go]: can throw in an `else if err !=nil { return fmt.Errorf("Unexpected error: %s", err) }`

File: helper.go [hunk 4]
```diff
@@ -2,9 +2,12 @@ package tftest
 
 import (
 	"fmt"
+	getter "github.com/hashicorp/go-getter"
```
Comment (radeksimko) [inline helper.go]: Nitpick: It's a convention to keep external deps separate from the stdlib ones.

File: helper.go [hunk 5]
```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
...
+	}
+
+	zipDecompressor := new(getter.ZipDecompressor)
+
+	for _, provider := range providers {
+		providerFilename := provider.Name()
+		providerNameExt := filepath.Ext(providerFilename)
+		providerName := strings.TrimSuffix(providerFilename, providerNameExt)
+		providerPath := filepath.Join(auxiliaryProviderDir, providerName)
+		providerSymlinkPath := filepath.Join(pluginDir, providerName)
```
Comment (radeksimko) [inline helper.go]: Nitpick: Nearly all variables in the for loop have `provider` prefix and I suppose it's somewhat obvious we are dealing with providers here - how do you feel about omitting that prefix?

File: helper.go [hunk 6]
```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
...
+		}
+
+		// if filename ends in .zip, assume it is a zip and extract it
+		// otherwise assume it is a provider binary
+		if providerNameExt == ".zip" {
+			_, err = os.Stat(providerPath)
+			if err == nil {
+				// assume we have already extracted this provider zip
+			} else {
+				zipDecompressor.Decompress(providerPath, filepath.Join(auxiliaryProviderDir, providerFilename), false)
```
Comment (radeksimko) [inline helper.go]: Are we intentionally ignoring errors here? What if the ZIP file is corrupted, password protected or something like that?

File: working_dir.go [hunk 7]
```diff
@@ -116,7 +116,6 @@ func (wd *WorkingDir) RequireClearPlan(t TestControl) {
 func (wd *WorkingDir) init(pluginDir string) error {
 	args := []string{"init"}
 	args = append(args, wd.baseArgs...)
-	args = append(args, "-plugin-dir="+pluginDir, wd.configDir)
```
Comment (radeksimko) [inline working_dir.go:119]: Was this removal intentional? 👀

Comment (kmoe) [reply working_dir.go:119]: Yes, this removal is the key change in this PR, allowing Terraform CLI to use the normal plugin discovery mechanism. Currently, if `-plugin-dir` is set, then Terraform CLI will search _only_ that directory for plugins.

An earlier solution involved changing this behaviour when the `-get-plugins` flag is explicitly set to `true`. Please see https://github.com/hashicorp/terraform/pull/23982 for a discussion of how these flags work.

If we're happy with this removal then the Core PR can be closed, as it is not a trivial change in functionality.

**Comment by kmoe** [comment]
Thanks for the review comments - ready for re-review.

File: helper.go [hunk 8]
```diff
@@ -107,6 +121,81 @@ func InitHelper(config *Config) (*Helper, error) {
...
+			continue
+		}
+
+		// if filename ends in .zip, assume it is a zip and extract it
+		// otherwise assume it is a provider binary
+		if filenameExt == ".zip" {
+			_, err = os.Stat(path)
+			if os.IsNotExist(err) {
+				zipDecompressor.Decompress(path, filepath.Join(auxiliaryProviderDir, filename), false)
+			} else if err == nil {
```
Comment (appilon) [inline helper.go]: I think this was meant to be `err != nil`

**Review by appilon — CHANGES_REQUESTED** [review]

File: helper.go (hunk 8 above)
Comment (kmoe) [reply helper.go]: Thank you

**Review by appilon — APPROVED** [review]
